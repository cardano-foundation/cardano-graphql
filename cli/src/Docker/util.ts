import { QuestionCollection } from 'inquirer'
import { DockerComposeStack, StatefulService } from './index'

export function selectServicesFromStack (stack: DockerComposeStack, multi = true): QuestionCollection {
  const label = multi === true ? 'services' : 'service'
  return {
    name: label,
    message: `Select ${label}`,
    type: multi === true ? 'checkbox' : 'list',
    choices: [...stack.statefulServices.values()].map(service => {
      return {
        name: service.name,
        value: service
      }
    }),
    validate (input) {
      if (input.length === 0) {
        throw new Error('At least one service must be selected')
      }
      return true
    }
  }
}

export function servicesFromNameArg (stack: DockerComposeStack, names: string): StatefulService[] {
  return names.split(',').map((name: string) => {
    if (!stack.statefulServices.has(name)) {
      throw new Error(`Unknown service ${name}`)
    }
    return stack.statefulServices.get(name)
  })
}
