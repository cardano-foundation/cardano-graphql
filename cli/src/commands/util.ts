import { DockerComposeStack } from '../docker'

export function serviceChoices (stack: DockerComposeStack) {
  return [...stack.statefulServices.values()].map(service => {
    return { name: service.name, value: service }
  })
}

export function serviceFromName (stack: DockerComposeStack, name: string) {
  if (!stack.statefulServices.has(name)) {
    throw new Error(`Unknown service ${name}`)
  }
  return stack.statefulServices.get(name)
}

export function servicesFromNames (stack: DockerComposeStack, names: string) {
  return names.split(',').map((name: string) => serviceFromName(stack, name))
}
