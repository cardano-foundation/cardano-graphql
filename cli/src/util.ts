
export function ensureValue (input: string) {
  return !input ? 'Cannot be blank' : true
}
