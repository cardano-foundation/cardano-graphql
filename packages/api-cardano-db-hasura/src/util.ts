import { ConnectionConfig, createInteractionContext } from '@cardano-ogmios/client'
import { Logger } from 'ts-log'

export const createInteractionContextWithLogger = (connection: ConnectionConfig, logger: Logger, module: string, onClose?: () => Promise<void>) =>
  createInteractionContext(
    (error) => {
      logger.error({ module, error }, error.message)
    },
    async (code, reason) => {
      if (code === 1000) {
        logger.info({ module, code }, reason)
      } else {
        logger.error({ module, code }, 'Connection closed')
        await onClose?.()
      }
    },
    {
      connection
    }
  )
