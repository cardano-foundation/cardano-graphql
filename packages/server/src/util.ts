import { Application } from 'express'
import http from 'http'

export function listenPromise (app: Application, port: number, hostname?: string): Promise<http.Server> {
  return new Promise(function (resolve, reject) {
    const server: http.Server = app.listen(port, hostname, () => resolve(server))
    server.on('error', reject)
  })
}
