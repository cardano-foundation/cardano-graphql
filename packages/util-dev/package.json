{
  "name": "@cardano-graphql/util-dev",
  "version": "1.0.0",
  "description": "Common development utilities",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -b ./src",
    "cleanup": "shx rm -rf dist node_modules",
    "lint": "eslint --ignore-path ../../.eslintignore \"**/*.ts\"",
    "prepack": "yarn build",
    "test": "shx echo No tests in this package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/input-output-hk/cardano-graphql.git"
  },
  "author": "Rhys Bartels-Waller",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/input-output-hk/cardano-graphql/issues"
  },
  "homepage": "https://github.com/input-output-hk/cardano-graphql/blob/master/packages/util-dev/README.md",
  "devDependencies": {
    "@cardano-graphql/util": "1.0.0",
    "shx": "^0.3.2"
  },
  "directories": {
    "lib": "src",
    "test": "test"
  },
  "files": [
    "dist",
    "LICENSE",
    "NOTICE",
    "README.md"
  ]
}
