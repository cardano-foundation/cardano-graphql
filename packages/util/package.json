{
  "name": "@cardano-graphql/util",
  "version": "1.0.0",
  "description": "Common utilities",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -b ./src",
    "cleanup": "shx rm -rf dist node_modules",
    "lint": "eslint --ignore-path ../../.eslintignore \"**/*.ts\"",
    "prepack": "yarn build",
    "test": "shx echo No tests in this package"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/input-output-hk/cardano-graphql.git"
  },
  "author": "Rhys Bartels-Waller",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/input-output-hk/cardano-graphql/issues"
  },
  "homepage": "https://github.com/input-output-hk/cardano-graphql/blob/master/packages/util/README.md",
  "dependencies": {
    "graphql": "14.5.8",
    "graphql-bigint": "^1.0.0",
    "graphql-tag": "^2.10.3",
    "p-retry": "^4.2.0"
  },
  "devDependencies": {
    "@types/node": "^14.0.13",
    "shx": "^0.3.2",
    "typescript": "^3.9.5"
  },
  "directories": {
    "lib": "src",
    "test": "test"
  },
  "files": [
    "dist",
    "LICENSE",
    "NOTICE",
    "README.md"
  ]
}
