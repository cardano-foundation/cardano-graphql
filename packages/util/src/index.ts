import { onFailedAttemptFor } from './onFailedAttemptFor'
import { loadQueryNode } from './queryNodeLoading'
import * as dateTimeValidation from './dateTimeValidation'
import * as scalars from './scalars'

export default {
  dateTimeValidation,
  onFailedAttemptFor,
  loadQueryNode,
  scalars
}
