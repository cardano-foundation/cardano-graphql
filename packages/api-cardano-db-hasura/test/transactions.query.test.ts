import BigNumber from 'bignumber.js'
import path from 'path'

import { DocumentNode } from 'graphql'
import util from '@cardano-graphql/util'
import { TestClient } from '@cardano-graphql/util-dev'
import { tx05ad8b, txe68043 } from './data_assertions'
import { buildClient } from './util'

function loadQueryNode (name: string): Promise<DocumentNode> {
  return util.loadQueryNode(path.resolve(__dirname, '..', 'src', 'example_queries', 'transactions'), name)
}

describe('transactions', () => {
  let client: TestClient
  beforeAll(async () => {
    client = await buildClient()
  }, 60000)

  it('Returns transactions by hashes', async () => {
    const result = await client.query({
      query: await loadQueryNode('transactionsByHashesOrderByFee'),
      variables: { hashes: [txe68043.basic.hash, tx05ad8b.basic.hash] }
    })
    expect(result.data.transactions.length).toBe(2)
    expect(result.data.transactions[0].inputs[0].sourceTxHash).toBe(txe68043.basic.sourceTxHash)
    expect(result.data.transactions[1].inputs[0].sourceTxHash).toBe(tx05ad8b.basic.sourceTxHash)
    expect(result.data.transactions[0].outputs[0].index).toBe(0)
    expect(result.data).toMatchSnapshot()
  })

  it('Can return ordered by block index', async () => {
    const result = await client.query({
      query: await loadQueryNode('orderedTransactionsInBlock'),
      variables: { blockNumber: 3979532 }
    })
    expect(result.data.transactions.length).toBe(8)
    expect(result.data.transactions[0].blockIndex).toBe(0)
    expect(result.data.transactions[1].blockIndex).toBe(1)
    expect(result.data.transactions[2].blockIndex).toBe(2)
    expect(result.data.transactions[7].blockIndex).toBe(7)
  })

  it('Can return aggregated data', async () => {
    const result = await client.query({
      query: await loadQueryNode('aggregateDataWithinTransaction'),
      variables: { hashes: [txe68043.aggregated.hash, tx05ad8b.aggregated.hash] }
    })
    expect(result.data.transactions.length).toBe(2)
    const { transactions: txs } = result.data
    expect(txs).toEqual([tx05ad8b.aggregated, txe68043.aggregated])
    const outputsPlusFee = new BigNumber(txs[1].outputs_aggregate.aggregate.sum.value).plus(txs[1].fee).toString()
    expect(txs[1].inputs_aggregate.aggregate.sum.value).toEqual(outputsPlusFee)
    expect(result.data).toMatchSnapshot()
  })
  it('Can return filtered aggregated data', async () => {
    const result = await client.query({
      query: await loadQueryNode('filteredAggregateDataWithinTransaction'),
      variables: { hash: txe68043.aggregated_filtered.hash }
    })
    expect(result.data).toMatchSnapshot()
  })
})
