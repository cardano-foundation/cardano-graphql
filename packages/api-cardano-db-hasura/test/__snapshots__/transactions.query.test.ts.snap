// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactions Can return aggregated data 1`] = `
Object {
  "transactions": Array [
    Object {
      "deposit": 0,
      "fee": 171246,
      "hash": "05ad8b467095e0886713a38231ab9fe84e4031a433a9400ebf70ec9415e20102",
      "inputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "4924799649906",
          },
          "count": "1",
          "max": Object {
            "value": "4924799649906",
          },
          "min": Object {
            "value": "4924799649906",
          },
          "sum": Object {
            "value": "4924799649906",
          },
        },
      },
      "outputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "2462399739330",
          },
          "count": "2",
          "max": Object {
            "value": "4824799478660",
          },
          "min": Object {
            "value": "100000000000",
          },
          "sum": Object {
            "value": "4924799478660",
          },
        },
      },
    },
    Object {
      "deposit": 0,
      "fee": 171070,
      "hash": "e680432562b3b71fe44ca4eb8e29cb181d3a0858b3e2a643a55f7513d901bcae",
      "inputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "768403000000",
          },
          "count": "1",
          "max": Object {
            "value": "768403000000",
          },
          "min": Object {
            "value": "768403000000",
          },
          "sum": Object {
            "value": "768403000000",
          },
        },
      },
      "outputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "384201414465",
          },
          "count": "2",
          "max": Object {
            "value": "768401828930",
          },
          "min": Object {
            "value": "1000000",
          },
          "sum": Object {
            "value": "768402828930",
          },
        },
      },
    },
  ],
}
`;

exports[`transactions Can return filtered aggregated data 1`] = `
Object {
  "transactions": Array [
    Object {
      "hash": "e680432562b3b71fe44ca4eb8e29cb181d3a0858b3e2a643a55f7513d901bcae",
      "inputs_aggregate": Object {
        "aggregate": Object {
          "count": "1",
        },
      },
      "outputs_aggregate": Object {
        "aggregate": Object {
          "count": "1",
        },
      },
    },
  ],
}
`;

exports[`transactions Returns transactions by hashes 1`] = `
Object {
  "transactions": Array [
    Object {
      "block": Object {
        "number": 29021,
      },
      "blockIndex": 0,
      "deposit": 0,
      "fee": 171246,
      "hash": "05ad8b467095e0886713a38231ab9fe84e4031a433a9400ebf70ec9415e20102",
      "inputs": Array [
        Object {
          "address": "DdzFFzCqrht5ExAdoZVExXoZTpoMYGKxva3thvvvHapsLZQzSX3kCqwgqi5NSM2oUtHYYDqsSnvSGbqkarB6cSgDZohUhLWZ9KFdDWsa",
          "sourceTxHash": "f6aa5241cdabb2ca461d1644d56023033b6d32b70afb7a48b224ec3d82ae4a4f",
          "sourceTxIndex": 0,
          "value": "4924799649906",
        },
      ],
      "outputs": Array [
        Object {
          "address": "DdzFFzCqrhsxVv8Sjuj35Zkw42T9WjwXa6DasrNknm5M3bPVDnWJWwhX5NymoZncUArTfFYZ9sYGaA9htKSNYoK5Vkzx2vUtunAFt3eE",
          "index": 0,
          "value": "4824799478660",
        },
        Object {
          "address": "DdzFFzCqrhskotfhVwhLvNFaVGpA6C4yR9DXe56oEL4Ewmze51f1uQsc1cQb8qUyqgzjUPBgFZiVbuQu7BaXrQkouyvzjYjLqfJpKG5s",
          "index": 1,
          "value": "100000000000",
        },
      ],
      "size": 220,
      "totalOutput": "4924799478660",
    },
    Object {
      "block": Object {
        "number": 29021,
      },
      "blockIndex": 1,
      "deposit": 0,
      "fee": 171070,
      "hash": "e680432562b3b71fe44ca4eb8e29cb181d3a0858b3e2a643a55f7513d901bcae",
      "inputs": Array [
        Object {
          "address": "DdzFFzCqrhsggyaAMAUTjGtjBr1CTp8tTcHYWbqtoyQBZcaYHM16rjbUDawTwoVaEPawAMPLJmpJVXHBNxZnTgmQzqAcNDe6XvMe5BkB",
          "sourceTxHash": "74a10f4a5de09a393ef8b8f749c50e6938aa4f57908bbdaaefdf5e814fed281a",
          "sourceTxIndex": 0,
          "value": "768403000000",
        },
      ],
      "outputs": Array [
        Object {
          "address": "DdzFFzCqrhsuz652nVpjktdtiV44uWJLHv83m61S33gzfB4TBx7SKp3DgM18fBJznMrbUdsEFEvXW4LYqVKFE9fjMgVhmJP2LBhUvEe8",
          "index": 0,
          "value": "768401828930",
        },
        Object {
          "address": "DdzFFzCqrhsgrGt7zHydkE4cKxjB7SgsdnMPUFxYEwBu1u4sGobxZL5Xo3eDycCrtSYAUyw9aAsZZxFpTdgur6MwYDYMwAZ4xNGKRD3o",
          "index": 1,
          "value": "1000000",
        },
      ],
      "size": 216,
      "totalOutput": "768402828930",
    },
  ],
}
`;
